Sources/DeepLearning/Layer.swift

```diff
@@ -47,16 +56,43 @@ public protocol Layer: Differentiable & KeyPathIterable
     /// The output type of the layer.
     associatedtype Output: Differentiable
 
-    /// Returns the output obtained from applying the layer to an input.
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameters
+    ///   - input: The input to the layer.
+    ///   - context: The contextual informance for the layer application, e.g. the current learning
+    ///     phase.
+    /// - Returns: The output.
     @differentiable(wrt: (self, input))
     func applied(to input: Input, in context: Context) -> Output
 }
 
-public extension Layer
-    where TangentVector : AdditiveArithmetic, CotangentVector : AdditiveArithmetic {
-    func valueWithPullback(at input: Input, in context: Context)
+public extension Layer {
+    @available(*, deprecated,
+               message: "Switch to 'applied(to:in:)' for training, or 'inferring(from:)' for inference")
+    func applied(to input: Input) -> Output {
+        return inferring(from: input)
+    }
+
+    /// Returns the inference output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: The inference output.
+    func inferring(from input: Input) -> Output {
```

Comment:
How about `inferred(from:)` in the past tense to be consistent with `applied(to:)`?

Reply:
That would not be a semantically correct name -- past participle verbs should only be used when `self` is an object as in "apply a layer to something" (hence `applied(to:)`). If you use `inferred(from:)` then it will read like "infer a layer from something", which is clearly incorrect.